hal-core/src/addr.rs

```diff
@@ -50,7 +74,20 @@ pub trait Address:
         self.align_down(align) == self
     }
 
-    fn as_ptr(&self) -> *const ();
+    /// Returns `true` if `self` is aligned on the alignment of the specified
+    /// type.
+    #[inline]
+    fn is_aligned_for<T>(self) -> bool {
+        self.is_aligned(core::mem::align_of::<T>())
+    }
+
+    /// # Panics
+    ///
+    /// - If `self` is not aligned for a `T`-typed value.
+    fn as_ptr<T>(self) -> *mut T {
+        assert!(self.is_aligned_for::<T>());
```

Comment:
I suppose we need this restriction for rust, even though there's no aligned read requirement on x64?

Reply:
that's correct, will add a note